src/server.js

```diff
@@ -162,46 +187,33 @@ class Server extends EventEmitter {
      * @param socket
      * @param head The first packet of the tunneling stream (may be empty)
      */
-    onConnect(request, socket, head) {
-        let handlerOpts;
-        this.prepareRequestHandling(request)
-            .then((result) => {
-                handlerOpts = result;
-                handlerOpts.srcHead = head;
-
-                const data = { request, source: socket, head, handlerOpts, server: this };
-
-                if (handlerOpts.upstreamProxyUrlParsed) {
-                    this.log(handlerOpts.id, 'Using HandlerTunnelChain');
-                    return chain(data);
-                }
+    async onConnect(request, socket, head) {
+        try {
+            const handlerOpts = await this.prepareRequestHandling(request);
+            handlerOpts.srcHead = head;
 
-                this.log(handlerOpts.id, 'Using HandlerTunnelDirect');
-                return direct(data);
-            })
-            .catch((err) => {
-                if (err.message === 'Username contains an invalid colon') {
-                    err = new RequestError('Invalid colon in username of upstream proxy credentials', 502);
-                } else if (err.message === '407 Proxy Authentication Required') {
-                    err = new RequestError('Invalid upstream proxy credentials', 502);
-                } else if (err.code === 'ENOTFOUND') {
-                    if (err.proxy) {
-                        err = new RequestError('Failed to connect to upstream proxy', 502);
-                    } else {
-                        err = new RequestError('Target website does not exist', 404);
-                    }
-                }
+            const data = { request, source: socket, head, handlerOpts, server: this };
 
-                this.failRequest(request, err, handlerOpts);
-            });
+            if (handlerOpts.upstreamProxyUrlParsed) {
+                this.log(socket.proxyChainId, 'Using HandlerTunnelChain');
+                return await chain(data);
+            }
+
+            this.log(socket.proxyChainId, 'Using HandlerTunnelDirect');
+            return await direct(data);
+        } catch (error) {
+            this.failRequest(request, this.normalizeHandlerError(error));
+        }
     }
 
     /**
      * Authenticates a new request and determines upstream proxy URL using the user function.
      * Returns a promise resolving to an object that can be passed to construcot of one of the HandlerXxx classes.
      * @param request
      */
-    prepareRequestHandling(request) {
+    async prepareRequestHandling(request) {
```

Comment:
this method is quite huge, would be good to split the implementation a bit

Reply:
Should we name https://github.com/apify/proxy-chain/blob/042b9dfa34ef0fbe3e0e7e031237e5befb018bf4/src/server.js#L282-L314 `callPrepareRequestFunction` or leave it as it is?

Reply:
Makes sense

Reply:
Makes sense to leave it or extract it?

Reply:
To extract it :)